feat: KeyStreams

xavierchanth opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Add an abstraction layer around the notificationService subscription which allows the app developer to observe a set of keys as a Stream. There are various types included:

- Single Element Stream (any generic non-collection type) [KeyStream\<T>]
- Iterable [IterableKeyStream\<T>]
- List [ListKeyStream\<T>]
- Set [SetKeyStream\<T>]
- Map [MapKeyStream<K,V>]

**- How I did it**

Local changes and e2e testing

**- How to verify it**

Read the mixin then two base files first when reviewing.
See the tests I wrote

**- Description for the changelog**
Feat: KeyStreams
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

xavierchanth committed:
feat: keyStream

xavierchanth committed:
feat: keyStream

xavierchanth committed:
chore: Documentation for KeyStream

xavierchanth committed:
chore: Documentation for KeyStream

xavierchanth commented:
@gkc this feature should be good to go now. I used templates/macros to handle the dart docs, do you know of any better solutions?

xavierchanth committed:
chore: set logging back to default in key_stream_test.dart

xavierchanth committed:
chore: set logging back to default in key_stream_test.dart

gkc commented:
Hi @XavierChanth - will take me a while to review this as there's a lot here, so I may not get to it until Tuesday (but will try to do sooner if I can)

xavierchanth commented:
> Hi @XavierChanth - will take me a while to review this as there's a lot here, so I may not get to it until Tuesday (but will try to do sooner if I can)

No worries! It is quite a lot, I tried my best to make the implementation as simple as possible. There are a lot of similar patterns which should make reading easier.

gkc reviewed (changes requested):
Very nice! LGTM in general

I have a few requests
1. Need tests verifying that this all works as intended in the event of switching between atsigns
2. Some documentation targeted at package users (maybe in key_stream.dart)  describing typical usages
3. Maybe some additional documentation clarifying or explaining behaviours when switching between atsigna
4. Some documentation targeted at package maintainers (in key_stream_mixin.dart) explaining how it hooks in with notifications

xavierchanth commented:
> Very nice! LGTM in general
> 
> I have a few requests
> 
> 1. Need tests verifying that this all works as intended in the event of switching between atsigns
> 2. Some documentation targeted at package users (maybe in key_stream.dart)  describing typical usages
> 3. Maybe some additional documentation clarifying or explaining behaviours when switching between atsigna
> 4. Some documentation targeted at package maintainers (in key_stream_mixin.dart) explaining how it hooks in with notifications

Noted, will work on these when I get the chance.

gkc committed:
Cosmetic logging changes in sync_service_impl.dart and sync_util.dart

xavierchanth committed:
fix: Add support for atsign change and test against it

xavierchanth committed:
fix: Add support for atsign change and test against it

xavierchanth committed:
Merge branch 'trunk' into key_stream

xavierchanth committed:
fix: key stream tests

xavierchanth committed:
fix: key stream tests

xavierchanth commented on at_client/lib/src/key_stream/key_stream_mixin.dart:
```diff
@@ -0,0 +1,419 @@
+  static final AtSignLogger _logger = AtSignLogger('KeyStream');
+
+  @visibleForTesting
+  static bool disposeOnAtsignChange = true;
```
@gkc there was in fact was an instance where we switch atsigns. I've added this flag to check whether we should listen to the atsign change.

gkc replied:
👍👍👍

xavierchanth committed:
fix: key stream tests

xavierchanth committed:
fix: key stream tests

sitaram-kalluri committed:
feat: encrypt notify text data and add unit tests

sitaram-kalluri committed:
feat: encrypt notify text data and add unit tests

xavierchanth committed:
chore: dart format

xavierchanth committed:
chore: dart format

xavierchanth committed:
chore: update documentation

xavierchanth committed:
chore: update documentation

xavierchanth committed:
chore: update documentation

xavierchanth committed:
chore: update documentation

xavierchanth committed:
chore: update documentation

xavierchanth committed:
chore: update documentation

xavierchanth commented:
@gkc let me know what you think, I've updated the documentation.

gkc reviewed (approved):
LGTM - really good work.

gkc committed:
Merge remote-tracking branch 'origin/trunk' into trunk

xavierchanth committed:
fix: Close codefence in dart annotation

xavierchanth committed:
fix: Close codefence in dart annotation

srieteja committed:
first commit

srieteja committed:
at_lookup dependency override

srieteja committed:
added new fields to at_client_preference.dart

srieteja committed:
passing additional fields to AtLookupImpl const

srieteja committed:
using dependency override for at_lookup

srieteja committed:
using dependency override for at_lookup in e2e tests

srieteja committed:
Merge branch 'trunk' into sec_context

purnimavenkatasubbu committed:
Added test for bypassCache

purnimavenkatasubbu committed:
added wait

xavierchanth committed:
fix(key_stream): add onError handler

xavierchanth committed:
fix(key_stream): add onError handler

xavierchanth committed:
refactor(key_stream): Remove collection impl classes

xavierchanth committed:
refactor(key_stream): Remove collection impl classes

xavierchanth committed:
chore: add new line to EOF

xavierchanth committed:
chore: add new line to EOF

xavierchanth committed:
Merge branch 'trunk' into key_stream

xavierchanth committed:
fix(key_stream): type casting for onError function

xavierchanth committed:
fix(key_stream): type casting for onError function

xavierchanth committed:
Merge branch 'key_stream' of github.com:atsign-foundation/at_client_sdk into key_stream

xavierchanth committed:
fix(key_stream): onError fallback value

xavierchanth committed:
fix(key_stream): onError fallback value

xavierchanth commented:
@gkc requesting another review for:

- Added onError function to catch exceptions from `atClient.get` calls.
- For each collection, removed `<collection>_key_stream_impl.dart` and merged with `<collection>_key_stream.dart`.
    (i.e. removed an extra and unnecessary layer)

gkc committed:
Merge remote-tracking branch 'origin/trunk' into trunk

srieteja committed:
modifications to unit tests

srieteja committed:
Merge remote-tracking branch 'origin/sec_context' into sec_context

srieteja committed:
at_lookup dependency override for func-tests

gkc reviewed (approved):
LGTM. There is a functional test failing; although it seems unrelated, plz take a look

sitaram-kalluri committed:
fix: skip reserved keys from conflict resolution

sitaram-kalluri committed:
fix: pull sync cached key issue also

gkc commented:
@kalluriramkumar was the failing test here due to the same issue which you uncovered?

sitaram-kalluri committed:
fix: pull sync cached key issue also

sitaram-kalluri committed:
Merge pull request #603 from atsign-foundation/fix_sync_conflict_bug

fix: skip reserved keys from conflict resolution

xavierchanth commented:
> @kalluriramkumar was the failing test here due to the same issue which you uncovered?

@kalluriramkumar if it is, please merge.

purnimavenkatasubbu committed:
Merge branch 'trunk' of https://github.com/atsign-foundation/at_client_sdk into test_bypasscache

purnimavenkatasubbu committed:
added timeouts

srieteja committed:
Merge branch 'trunk' into sec_context

gkc committed:
Merge remote-tracking branch 'origin/trunk' into trunk

sitaram-kalluri committed:
fix: set autoNotify to true at end of test.

sitaram-kalluri committed:
fix: set autoNotify to true at end of test.

murali-shris committed:
Merge pull request #598 from atsign-foundation/test_bypasscache

feat: Added test for bypassCache

murali-shris committed:
publish at client 3.0.30

murali-shris committed:
modified change log

murali-shris committed:
Merge pull request #613 from atsign-foundation/publish_at_client

chore: publish at client 3.0.30

srieteja committed:
dependency version upgrades

srieteja committed:
changelog updated

srieteja committed:
Merge branch 'trunk' into sec_context

murali-shris committed:
Merge pull request #614 from atsign-foundation/srieteja-patch-3

chore:upgrading at_client_mobile dependency versions

gkc committed:
Merge remote-tracking branch 'origin/trunk' into trunk

gkc committed:
feat: remove network availability check from AtClientImpl.notifyChange()

gkc committed:
feat: remove network availability check from AtClientImpl.notifyChange()
feat: add optional `onSentToSecondary` callback function to NotificationService.notify(...) method signature

gkc committed:
feat: add optional 'waitForFinalDeliveryStatus' parameter to NotificationService.notify(...) and added explanatory documentation. In brief: currently we have two options in calling notify - either (1) completely asyncrhonously or (2) completely synchronously including getting the final delivery status, which could be a very long time. This parameter allows you to call the method synchronously using await but it will complete very rapidly i.e. once the notification has been sent to this client's secondary server, with the polling to check final delivery status continuing asynchronously.
feat: add optional 'checkForFinalDeliveryStatus' parameter so that if you wish you can not bother doing the polling for final delivery status here.

gkc committed:
* Changed an internal method name for readability
* Changed the logic which polls for delivery status so that it waits for only 500 milliseconds before first poll, and then waits 2 seconds between polls thereafter

gkc committed:
* chore: little but of de-linting in notification_service_impl.dart

sitaram-kalluri committed:
feat: Introduce notification_response_transformer.dart

sitaram-kalluri committed:
feat: Introduce notification_response_transformer.dart

sitaram-kalluri committed:
feat: Introduce notification_response_transformer.dart

sitaram-kalluri committed:
feat: Introduce notification_response_transformer.dart

gkc committed:
Merge pull request #615 from atsign-foundation/RemoveNetworkCheckFrom_AtClientImpl_notify

srieteja committed:
passing instance of cacheableSecondaryFinder to atLookup

srieteja committed:
change CacheableSecondaryAddressFinder to SecondaryAddressFinder

purnimavenkatasubbu committed:
fix for refraining sync

srieteja committed:
dependency override for at_lookup in func_tests and e2e tests

sitaram-kalluri committed:
fix: remove network checking from notification_service_impl.dart

sitaram-kalluri committed:
fix: remove network checking from notification_service_impl.dart

sitaram-kalluri committed:
Merge remote-tracking branch 'origin/encrypt_notify_text' into encrypt_notify_text

# Conflicts:
#	at_client/lib/src/service/notification_service_impl.dart
#	at_client/test/notification_service_test.dart

srieteja committed:
setting decryptPackets to true

srieteja committed:
providing a directory to save tlsKeysFile in func_tests

srieteja committed:
adding a cert to func_tests to test SecureSocketutil

sitaram-kalluri committed:
fix: Added e2e tests. Update version and CHANGELOG.md

srieteja committed:
removing dependency overrides for at_lookup

srieteja committed:
removing dependency overrides for at_lookup

srieteja committed:
refactoring code to use SecureSocketConfig

srieteja committed:
refactoring code to use SecureSocketConfig

sitaram-kalluri committed:
fix: remove NetworkCheckConnectivity from network_util.dart

sitaram-kalluri committed:
fix: Add timeout to notify_test.dart

sitaram-kalluri committed:
fix: fail e2e tests

sitaram-kalluri committed:
fix: Update the CHANGELOG.md

sitaram-kalluri committed:
Merge pull request #611 from atsign-foundation/encrypt_notify_text

feat: encrypt notify text data and add unit tests

srieteja committed:
Merge branch 'trunk' into sec_context

srieteja committed:
Merge branch 'trunk' into multiple_calls_find_secondary

srieteja committed:
Merge pull request #590 from atsign-foundation/sec_context

feat:support for optional decryptPackets parameter

purnimavenkatasubbu committed:
fixed failing test

purnimavenkatasubbu committed:
dependency update

srieteja committed:
Merge branch 'trunk' into multiple_calls_find_secondary

purnimavenkatasubbu committed:
updated dependencies

srieteja committed:
correcting syntax error

purnimavenkatasubbu committed:
updated published version

purnimavenkatasubbu committed:
Merge pull request #619 from atsign-foundation/issue_596

fix: do not sync statsNotificationID from client to server

srieteja committed:
Merge pull request #618 from atsign-foundation/multiple_calls_find_secondary

fix:instance of cacheableSecondaryFinder to atLookup

xavierchanth committed:
Merge branch 'key_stream' of github.com:atsign-foundation/at_client_sdk into key_stream

xavierchanth committed:
fix: duplicate export
